### PR TITLE
refactor(parser): Invert event processing

### PR DESCRIPTION
**Pull request by fb55**
See #403 for details.

The parser currently first checks the insertion mode, and then the token type. By inverting this (checking first the token type, then the insertion mode), we prepare the parser to accept the events from #404.

This change also allowed foreign content checks to be greatly simplified.

### CODE REVIEW

**Comment by fb55** [comment]
One thing I am currently unhappy about is the `tokenizer.inForeignNode` and `currentNotInHTML` naming inside of the parser.

- `currentNotInHTML` indicates that the current node is not an element in the HTML namespace.
- `tokenizer.inForeignNode` (fka. `tokenizer.allowCDATA`) indicates that the current node is not an element in the HTML namespace, and that it is not an integration point for either MathML or HTML. (See https://html.spec.whatwg.org/#tree-construction)

If someone can come up with better names, I'm am more than happy to take them.

File: packages/parse5/lib/tokenizer/index.ts [hunk 1]
```diff
@@ -209,7 +209,13 @@ export class Tokenizer {
 
     private tokenQueue: Token[] = [];
 
-    public allowCDATA = false;
+    /**
+     * Indicates that the current adjusted node exists, is not an element in the HTML namespace,
+     * and that it is not an integration point for either MathML or HTML.
+     *
+     * @see {@link https://html.spec.whatwg.org/#tree-construction}
```
Comment (wooorm) [inline packages/parse5/lib/tokenizer/index.ts:216]: link to multipage perhaps, otherwise it’s too slow?

**Review by wooorm — COMMENTED** [review]
Isn’t this a deviation from how the spec is written? Seems like that’ll cause future problems.

**Comment by fb55** [comment]
> Isn’t this a deviation from how the spec is written? Seems like that’ll cause future problems.

I don't think this will make maintenance noticeably harder. parse5 already has a split for specific token handlers that doesn't exist in the spec. The splitting point was moved up by a step, which shouldn't make it harder to map the spec to the code.

**Pull request merged**